File: Server/src/models/PokemonStats.ts
import mongoose, { Document, Schema, Types } from 'mongoose';

export interface IStatBlock {
  hp: number;
  atk: number;
  def: number;
  spatk: number;
  spdef: number;
  spd: number;
}

export interface IPokemonStats extends Document {
  pokemon_instance_id: Types.ObjectId;
  ivs: IStatBlock;
  evs: IStatBlock;
}

const StatBlockSchema = new Schema({
  hp: { type: Number, required: true, min: 0, max: 255 },
  atk: { type: Number, required: true, min: 0, max: 255 },
  def: { type: Number, required: true, min: 0, max: 255 },
  spatk: { type: Number, required: true, min: 0, max: 255 },
  spdef: { type: Number, required: true, min: 0, max: 255 },
  spd: { type: Number, required: true, min: 0, max: 255 }
}, { _id: false });

const PokemonStatsSchema: Schema = new Schema({
  pokemon_instance_id: { type: Schema.Types.ObjectId, ref: 'PokemonInstance', required: true, unique: true },
  ivs: { type: StatBlockSchema, required: true },
  evs: { type: StatBlockSchema, required: true }
});

export const PokemonStatsModel = mongoose.model<IPokemonStats>('PokemonStats', PokemonStatsSchema);

File: Server/Models/PokemonStats.cs
using MongoDB.Bson;
using MongoDB.Bson.Serialization.Attributes;

namespace PokemonMMO.Models;

/// <summary>
/// Individual Values (IVs) and Effort Values (EVs) for a pokemon instance.
/// </summary>
public class PokemonStats
{
    [BsonId]
    [BsonRepresentation(BsonType.ObjectId)]
    public string Id { get; set; } = null!;

    [BsonElement("pokemon_instance_id")]
    [BsonRepresentation(BsonType.ObjectId)]
    public string PokemonInstanceId { get; set; } = null!;

    [BsonElement("ivs")]
    public StatBlock Ivs { get; set; } = new();

    [BsonElement("evs")]
    public StatBlock Evs { get; set; } = new();
}

/// <summary>
/// Six core stat values (HP, ATK, DEF, SpATK, SpDEF, SPD).
/// Range: 0–31 for IVs, 0–255 for EVs.
/// </summary>
public class StatBlock
{
    [BsonElement("hp")]    public int Hp    { get; set; }
    [BsonElement("atk")]   public int Atk   { get; set; }
    [BsonElement("def")]   public int Def   { get; set; }
    [BsonElement("spatk")] public int SpAtk { get; set; }
    [BsonElement("spdef")] public int SpDef { get; set; }
    [BsonElement("spd")]   public int Spd   { get; set; }
}
